# Ejercicio 4 

Autora: Blanca Cano Camarero   
Correo: BlancaCano@correo.ugr.es

Dado el número n

In [2]:
n = 75577399
is_prime(n)

True

### Apartado i
Factoriza $n-1$ aplicando el método $\rho$ de Polard ¿Cuántas iteraciones necesitas?

In [3]:
def Rho_de_Polard(n :Integer, f) -> Integer:
    '''Devuelve un factor de p
    o n si es un fracaso
    '''
    x = 1
    y = 1
    d = 1
    while d == 1:
        x = f(x)
        y = f(f(y))
        d = gcd(abs(x-y),n)
    return d

def Verbose_Rho_de_Polard(n:Integer, f) -> Integer:
    '''Devuelve un factor de p
    o n si es un fracaso
    '''
    mf = lambda x: f(x)% n
    print(f"Cálculo de Rho de Polar para n={n}")
    x = 1
    y = 1
    d = 1
    paso = 0
    print(f"Paso {paso}   x={x}    y={y}")
    while d == 1:
        paso += 1
        x = mf(x)
        y = mf(mf(y))
        d = gcd(abs(x-y),n)
        print(f"Paso {paso}    x={x}    y={y}    mcd={d}")
    return d
f = lambda x : x^2 +1 

## Versión de sacar todos los factores de manera manual 
def SacaFactores(numero_factores, n, f ):
    factores = []
    m = n
    hay_factores = True 
    while numero_factores >= 0 and hay_factores:
        numero_factores -= 1
        d = Verbose_Rho_de_Polard(m,f)
        if d != m:
            factores.append(d)
            m = m // d
            print(f"Se tiene que n = multiplicar los elementos de {factores} y {m}\n")
        else:
            factores.append(m)
            print('='*60)
            hay_factores = False
    return factores
factores = SacaFactores(5, n-1,f )
print(f"Los factores de {n-1} son {factores}")


Cálculo de Rho de Polar para n=75577398
Paso 0   x=1    y=1
Paso 1    x=2    y=5    mcd=3
Se tiene que n = multiplicar los elementos de [3] y 25192466

Cálculo de Rho de Polar para n=25192466
Paso 0   x=1    y=1
Paso 1    x=2    y=5    mcd=1
Paso 2    x=5    y=677    mcd=2
Se tiene que n = multiplicar los elementos de [3, 2] y 12596233

Cálculo de Rho de Polar para n=12596233
Paso 0   x=1    y=1
Paso 1    x=2    y=5    mcd=1
Paso 2    x=5    y=677    mcd=1
Paso 3    x=26    y=11607393    mcd=1
Paso 4    x=677    y=4763383    mcd=13
Se tiene que n = multiplicar los elementos de [3, 2, 13] y 968941

Cálculo de Rho de Polar para n=968941
Paso 0   x=1    y=1
Paso 1    x=2    y=5    mcd=1
Paso 2    x=5    y=677    mcd=1
Paso 3    x=26    y=949042    mcd=1
Paso 4    x=677    y=887619    mcd=1
Paso 5    x=458330    y=256330    mcd=1
Paso 6    x=949042    y=276002    mcd=1
Paso 7    x=642274    y=952119    mcd=1
Paso 8    x=887619    y=166241    mcd=1
Paso 9    x=245360    y=528585    mcd=1
Pa

### Apartado ii

Vamos a comprobar que son factores primos con el Test de Lucas-Lehmer.

Que recordamos que era si cumple que: 
Existe un entero $a$ tal que: 
1. $a^{n-1} \equiv 1 \mod n$
2. $a^{(n-1)/q} \not \equiv 1 \mod n$ para $q$ divisor de $n-1.$

Entonces, $n$ es primo.

In [4]:
def Test_Lucas_Lehmer(n, bases):
    m = n-1
    for a in bases:
        # cumple condición 1
        if mod(a^m, n) == 1:
            condicion_dos = True
            for i in range (2,m):
                # si es divisor 
                if m % i == 0:
                    if (int(a**(m/i))% n) == 1:
                        condicion_dos =  False
                        break
            if condicion_dos == True:
                return a
    return 0

for p in factores[3:]:
    resultado = Test_Lucas_Lehmer(p,[2,3,5,7,11,13])
    mensaje = (
        f"p={p} es primo por el test Lucas-Lehmer con a = {resultado}",
        "No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {p}"
    )[resultado == 0] 
    print(mensaje)

p=359 es primo por el test Lucas-Lehmer con a = 7
p=2699 es primo por el test Lucas-Lehmer con a = 2


#### Conclusión 

Acabamos de probar, ya sea por ser $2,3$ y $13$ primos ampliamente conocidos o por el Test de Lucas Lehmer que los factores $[3, 2, 13, 359, 2699]$ del apartado primero son factores primos.  

### Apartado tercero
Aplica Lucas-Lehmer para encontrar un certificado de primalidad de $n$. 
Vamos a encontrar un testigo de primalidad. 


In [5]:
resultado = Test_Lucas_Lehmer(n,[2699])
mensaje = (
    f"p={n} es primo por el test Lucas-Lehmer con a = {resultado}",
    f"No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {n}"
)[resultado == 0] 
print(mensaje)  

No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = 75577399


In [10]:
def BuscaTestigo(n, factores, lista):
    m = n-1
    for i in lista:
        if pow(i,m, n) == 1:
            condicion_testigo = True 
            for f in factores:
                if pow(i,m//f, n) == 1:
                    condicion_testigo = False 
                    break
            if condicion_testigo == True:
                return i
    return 0

testigo = BuscaTestigo(n, factores, [2,3,5,7,11,13,15])
mensaje = (
    f"{testigo} es un testigo de la primalidad de n={n}",
    f"No se ha encontrado un testigo que demuestre la primalidad de n= {n}"
)[testigo == 0] 
print(mensaje)


resultado = Test_Lucas_Lehmer(n,[testigo])
mensaje = (
    f"p={n} es primo por el test Lucas-Lehmer con a = {resultado}",
    f"No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {n}"
)[resultado == 0] 
print(mensaje)  
